src/framework/MOM_array_transform.F90

```diff
@@ -0,0 +1,358 @@
+!! +90deg: B(i,j) = A(n-j,i)
+!!                = transpose, then row reverse
+!! 180deg: B(i,j) = A(m-i,n-j)
+!!                = row reversal + column reversal
+!! -90deg: B(i,j) = A(j,m-i)
+!!                = row reverse, then transpose
+!!
+!! 90 degree rotations change the shape of the field, and are handled
+!! separately from 180 degree rotations.
+
+module MOM_array_transform
+
+implicit none
+
+private
+public rotate_array
+public rotate_array_pair
+public rotate_vector
+public allocate_rotated_array
+
+
+!> Rotate the elements of an array to the rotated set of indices.
+!! Rotation is applied across the first and second axes of the array.
+interface rotate_array
+  module procedure rotate_array_real_2d
+  module procedure rotate_array_real_3d
+  module procedure rotate_array_real_4d
+  module procedure rotate_array_integer
+  module procedure rotate_array_logical
+end interface rotate_array
+
+
+!> Rotate a pair of arrays which map to a rotated set of indices.
+!! Rotation is applied across the first and second axes of the array.
+!! This rotation should be applied when one field is mapped onto the other.
+!! For example, a tracer indexed along u or v face points will map from one
+!! to the other after a quarter turn, and back onto itself after a half turn.
+interface rotate_array_pair
+  module procedure rotate_array_pair_real_2d
+  module procedure rotate_array_pair_real_3d
+  module procedure rotate_array_pair_integer
+end interface rotate_array_pair
+
+
+!> Rotate an array pair representing the components of a vector.
+!! Rotation is applied across the first and second axes of the array.
+!! This rotation should be applied when the fields satisfy vector
+!! transformation rules.  For example, the u and v components of a velocity
+!! will map from one to the other for quarter turns, with a sign change in one
+!! component.  A half turn will map elements onto themselves with sign changes
+!! in both components.
+interface rotate_vector
+  module procedure rotate_vector_real_2d
+  module procedure rotate_vector_real_3d
+  module procedure rotate_vector_real_4d
+end interface rotate_vector
+
+
+!> Allocate an array based on the rotated index map of an unrotated reference
+!! array.
+interface allocate_rotated_array
+  module procedure allocate_rotated_array_real_2d
+  module procedure allocate_rotated_array_real_3d
+  module procedure allocate_rotated_array_real_4d
+  module procedure allocate_rotated_array_integer
+end interface allocate_rotated_array
+
+contains
+
+!> Rotate the elements of a 2d real array along first and second axes.
+subroutine rotate_array_real_2d(A_in, turns, A)
+  real, intent(in) :: A_in(:,:) !< Unrotated array
+  integer, intent(in) :: turns  !< Number of quarter turns
+  real, intent(out) :: A(:,:)   !< Rotated array
+
+  integer :: m, n
+
+  m = size(A_in, 1)
+  n = size(A_in, 2)
+
+  select case (modulo(turns, 4))
+    case(0)
+      A = A_in
+    case(1)
+      A = transpose(A_in)
+      A = A(n:1:-1, :)
```

Comment:
The use of array syntax is generally discouraged in MOM6, as described in the MOM6 style guide (see https://github.com/NOAA-GFDL/MOM6/wiki/Code-style-guide#array-syntax). In this case, there is a certain compactness to this copy as it avoids the need for an explicit temporary array, but I think that it would be preferable to write this line as A(:,:) = A(n:1:-1, :) to make it more explicit that array syntax is being used.

Reply:
Agreed. Adding `A(:,:) = A(n:1:-1,:)` works better than adding `HI` to these interfaces and since these operations are over the entire array the syntax works well. It's the partial loop ranges that are indecipherable using array syntax.